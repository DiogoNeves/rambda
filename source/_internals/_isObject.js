import { _isArray } from "./_isArray";

export function _isObject (item){
  if (!item) return false;

  return typeof item === 'object' && !_isArray(item);
}
