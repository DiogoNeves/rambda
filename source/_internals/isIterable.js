import {type} from '../type'

export function _isIterable(input) {
  return ['Object', 'Array'].includes(type(input))
}
