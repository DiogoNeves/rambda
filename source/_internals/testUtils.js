import combinate from 'combinate'
import { equals, filter, map, omit, switcher, type } from 'rambdax'

const omitOk = omit('ok')

const PENDING = 'PENDING'
const RESULTS_EQUAL = 'results are equal'
const RESULTS_MISMATCH = 'results are different'
const ERRORS_EQUAL = 'errors are equal'
const ERRORS_MISMATCH = 'errors are different'
const SHOULD_THROW = 'Rambda should throw'
const SHOULD_NOT_THROW = 'Rambda should not throw'

const MISSING = 'Missing error handle in parseError'

function parseError(err){
  const typeError = switcher(err)
    .is(x => x instanceof TypeError, 'TypeError')
    .is(x => x instanceof TypeError, 'TypeError')
    .is(x => x instanceof TypeError, 'TypeError')
    .default(MISSING)

  if (typeError === MISSING){
    throw new Error('typeError === MISSING')
  }

  return {
    message : err.message,
    type    : typeError,
    ok      : true,
  }
}

export function show(input){
  const typeInput = type(input)
  if ([ 'Promise', 'Async' ].includes(typeInput)){
    return ''
  }

  if (typeInput === 'Array'){
    if (input.length === 0) return '[]'

    return `[${ input.map(show).join(', ') }]`
  }

  if (typeInput === 'Object'){
    if (Object.keys(input).length === 0) return '{}'

    return JSON.stringify(map(show, input))
  }
  if ([ 'Boolean', 'Number', 'String' ].includes(typeInput)){
    return input
  }
  if ([ 'Null', 'Undefined' ].includes(typeInput)){
    return typeInput.toLowerCase()
  }

  return input.toString()
}

function executeSync(fn, inputs){
  let result = PENDING
  let error = { ok : false }
  try {
    result = fn(...inputs)
  } catch (e){
    error = parseError(e)
  }

  return {
    result,
    error,
  }
}

async function executeAsync(fn, inputs){
  let result = PENDING
  let error = { ok : false }
  try {
    result = await fn(...inputs)
  } catch (e){
    error = parseError(e)
  }

  return {
    result,
    error,
  }
}

export function profileMethod(
  firstInput,
  secondInput = undefined,
  thirdInput = undefined,
  fn
){
  const combinationsInput = filter(Boolean, {
    firstInput,
    secondInput,
    thirdInput,
  })
  const inputKeys = Object.keys(combinationsInput)
  const combinations = combinate(combinationsInput)

  combinations.forEach(combination => {
    const inputs = [
      combination.firstInput,
      combination.secondInput,
      combination.thirdInput,
    ].filter((_, i) => i < inputKeys.length)

    test(getTestTitle(...inputs), () => {
      const { result, error } = executeSync(fn, inputs)

      expect({
        result,
        error : error.ok ? omitOk(error) : PENDING,
        inputs,
      }).toMatchSnapshot()
    })
  })
}

export function profileMethodAsync({
  firstInput,
  secondInput = undefined,
  thirdInput = undefined,
  fn,
  callback,
}){
  afterAll(() => callback())
  const combinationsInput = filter(Boolean, {
    firstInput,
    secondInput,
    thirdInput,
  })
  const inputKeys = Object.keys(combinationsInput)
  const combinations = combinate(combinationsInput)

  combinations.forEach(combination => {
    const inputs = [
      combination.firstInput,
      combination.secondInput,
      combination.thirdInput,
    ].filter((_, i) => i < inputKeys.length)
    test(getTestTitle(...inputs), async () => {
      const { result, error } = await executeAsync(fn, inputs)

      expect({
        result,
        error : error.ok ? omitOk(error) : PENDING,
        inputs,
      }).toMatchSnapshot()
    })
  })
}

export function compareToRamda(fn, fnRamda){
  return (...inputs) => {
    const { result, error } = executeSync(fn, inputs)
    const { result: ramdaResult, error: ramdaError } = executeSync(fnRamda,
      inputs)

    const toReturn = {
      result,
      ramdaResult,
      ramdaError : ramdaError.ok ? omitOk(ramdaError) : PENDING,
      error      : error.ok ? omitOk(error) : PENDING,
    }

    if (result !== PENDING){
      if (ramdaError.ok){
        return {
          ...toReturn,
          ok    : false,
          label : SHOULD_THROW,
        }
      }

      if (equals(result, ramdaResult)){
        return {
          ...toReturn,
          ok    : true,
          label : RESULTS_EQUAL,
        }
      }

      return {
        ...toReturn,
        ok    : false,
        label : RESULTS_MISMATCH,
      }
    }

    if (equals(error, ramdaError)){
      return {
        ...toReturn,
        ok    : true,
        label : ERRORS_EQUAL,
      }
    }

    if (ramdaError.ok){
      return {
        ...toReturn,
        ok    : false,
        label : ERRORS_MISMATCH,
      }
    }

    return {
      ...toReturn,
      ok    : false,
      label : SHOULD_NOT_THROW,
    }
  }
}

export const getTestTitle = (...inputs) =>
  inputs.map(x => `${ type(x) } ${ show(x) }`).join(' | ')

export const compareCombinations = ({
  firstInput,
  secondInput = undefined,
  thirdInput = undefined,
  setCounter = () => {},
  setGlobalCounter = () => {},
  fn,
  fnRamda,
}) => {
  const combinationsInput = filter(Boolean, {
    firstInput,
    secondInput,
    thirdInput,
  })
  const inputKeys = Object.keys(combinationsInput)
  const combinations = combinate(combinationsInput)
  const compareOutputs = compareToRamda(fn, fnRamda)

  combinations.forEach(combination => {
    const inputs = [
      combination.firstInput,
      combination.secondInput,
      combination.thirdInput,
    ].filter((_, i) => i < inputKeys.length)

    test(getTestTitle(...inputs), () => {
      const compared = compareOutputs(...inputs)
      setGlobalCounter()
      if (!compared.ok){
        setCounter()
        expect({
          ...compared,
          inputs,
        }).toMatchSnapshot()
      }
    })
  })
}
