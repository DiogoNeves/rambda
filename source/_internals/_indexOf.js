import { equals } from '../equals'
import {_isObject} from './_isObject'

/**
 * Get first index of item in a list based on native `indexOf` or on `R.equals` if input is list of objects.
 */
export function _indexOf(valueToFind, list) {
  if (!_isObject(valueToFind)) return list.indexOf(valueToFind)

  let index = -1
  list.forEach((x, i) => {
    if (index > -1) return
    if (equals(x, valueToFind)) index = i
  })

  return index
}
