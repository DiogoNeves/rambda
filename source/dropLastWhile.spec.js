import { dropLastWhile as dropLastWhileRamda } from 'ramda'

import { compareCombinations } from './_internals/testUtils'
import { dropLastWhile } from './dropLastWhile'

const list = [ 1, 2, 3, 4, 5 ]
const str = 'foobar'

test('with list', () => {
  const result = dropLastWhile(x => x >= 3, list)
  expect(result).toEqual([ 1, 2 ])
})

test('with string', () => {
  const result = dropLastWhile(x => x !== 'b')(str)
  expect(result).toBe('foob')
})

test('with empty list', () => {
  expect(dropLastWhile(() => true, [])).toEqual([])
  expect(dropLastWhile(() => false, [])).toEqual([])
})

const possiblePredicates = [
  x => x > 2,
  x => x < 2,
  x => x < -2,
  x => x > 10,
  '',
  [],
  [ 1 ],
]

const possibleIterables = [
  list,
  [ {}, '1', 2 ],
  str,
  `${ str }${ str }`,
  /foo/g,
  Promise.resolve('foo'),
  2,
]

test.skip('foo', () => {
  const predicate = []
  const iterable = 2
  const a = dropLastWhile(predicate, iterable)
  const b = dropLastWhileRamda(predicate, iterable)
  console.log({ a })
  console.log({ b })
})

describe('brute force', () => {
  compareCombinations({
    fn          : dropLastWhile,
    fnRamda     : dropLastWhileRamda,
    firstInput  : possiblePredicates,
    secondInput : possibleIterables,
    callback    : errorsCounters => {
      expect(errorsCounters).toMatchInlineSnapshot(`
        Object {
          "ERRORS_DIFFERENT": 0,
          "ERRORS_MISMATCH": 12,
          "RESULTS_MISMATCH": 0,
          "SHOULD_NOT_THROW": 21,
          "SHOULD_THROW": 0,
        }
      `)
    },
  })
})
