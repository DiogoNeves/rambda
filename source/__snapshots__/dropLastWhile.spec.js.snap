// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`brute force Array [] | Array [{}, 1, 2] 1`] = `
Object {
  "error": Object {
    "message": "'predicate' is from wrong type object",
    "type": "Error",
  },
  "inputs": Array [
    Array [],
    Array [
      Object {},
      "1",
      2,
    ],
  ],
  "label": "errors are different",
  "ok": false,
  "ramdaError": Object {
    "message": "pred is not a function",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": "PENDING",
}
`;

exports[`brute force Array [] | Array [1, 2, 3, 4, 5] 1`] = `
Object {
  "error": Object {
    "message": "'predicate' is from wrong type object",
    "type": "Error",
  },
  "inputs": Array [
    Array [],
    Array [
      1,
      2,
      3,
      4,
      5,
    ],
  ],
  "label": "errors are different",
  "ok": false,
  "ramdaError": Object {
    "message": "pred is not a function",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": "PENDING",
}
`;

exports[`brute force Array [] | Number 2 1`] = `
Object {
  "error": Object {
    "message": "'predicate' is from wrong type object",
    "type": "Error",
  },
  "inputs": Array [
    Array [],
    2,
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": Array [],
  "result": "PENDING",
}
`;

exports[`brute force Array [] | Promise  1`] = `
Object {
  "error": Object {
    "message": "'predicate' is from wrong type object",
    "type": "Error",
  },
  "inputs": Array [
    Array [],
    Promise {},
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": Array [],
  "result": "PENDING",
}
`;

exports[`brute force Array [] | RegExp /foo/g 1`] = `
Object {
  "error": Object {
    "message": "'predicate' is from wrong type object",
    "type": "Error",
  },
  "inputs": Array [
    Array [],
    /foo/g,
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": Array [],
  "result": "PENDING",
}
`;

exports[`brute force Array [] | String foobar 1`] = `
Object {
  "error": Object {
    "message": "'predicate' is from wrong type object",
    "type": "Error",
  },
  "inputs": Array [
    Array [],
    "foobar",
  ],
  "label": "errors are different",
  "ok": false,
  "ramdaError": Object {
    "message": "pred is not a function",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": "PENDING",
}
`;

exports[`brute force Array [] | String foobarfoobar 1`] = `
Object {
  "error": Object {
    "message": "'predicate' is from wrong type object",
    "type": "Error",
  },
  "inputs": Array [
    Array [],
    "foobarfoobar",
  ],
  "label": "errors are different",
  "ok": false,
  "ramdaError": Object {
    "message": "pred is not a function",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": "PENDING",
}
`;

exports[`brute force Array [1] | Array [{}, 1, 2] 1`] = `
Object {
  "error": Object {
    "message": "'predicate' is from wrong type object",
    "type": "Error",
  },
  "inputs": Array [
    Array [
      1,
    ],
    Array [
      Object {},
      "1",
      2,
    ],
  ],
  "label": "errors are different",
  "ok": false,
  "ramdaError": Object {
    "message": "pred is not a function",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": "PENDING",
}
`;

exports[`brute force Array [1] | Array [1, 2, 3, 4, 5] 1`] = `
Object {
  "error": Object {
    "message": "'predicate' is from wrong type object",
    "type": "Error",
  },
  "inputs": Array [
    Array [
      1,
    ],
    Array [
      1,
      2,
      3,
      4,
      5,
    ],
  ],
  "label": "errors are different",
  "ok": false,
  "ramdaError": Object {
    "message": "pred is not a function",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": "PENDING",
}
`;

exports[`brute force Array [1] | Number 2 1`] = `
Object {
  "error": Object {
    "message": "'predicate' is from wrong type object",
    "type": "Error",
  },
  "inputs": Array [
    Array [
      1,
    ],
    2,
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": Array [],
  "result": "PENDING",
}
`;

exports[`brute force Array [1] | Promise  1`] = `
Object {
  "error": Object {
    "message": "'predicate' is from wrong type object",
    "type": "Error",
  },
  "inputs": Array [
    Array [
      1,
    ],
    Promise {},
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": Array [],
  "result": "PENDING",
}
`;

exports[`brute force Array [1] | RegExp /foo/g 1`] = `
Object {
  "error": Object {
    "message": "'predicate' is from wrong type object",
    "type": "Error",
  },
  "inputs": Array [
    Array [
      1,
    ],
    /foo/g,
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": Array [],
  "result": "PENDING",
}
`;

exports[`brute force Array [1] | String foobar 1`] = `
Object {
  "error": Object {
    "message": "'predicate' is from wrong type object",
    "type": "Error",
  },
  "inputs": Array [
    Array [
      1,
    ],
    "foobar",
  ],
  "label": "errors are different",
  "ok": false,
  "ramdaError": Object {
    "message": "pred is not a function",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": "PENDING",
}
`;

exports[`brute force Array [1] | String foobarfoobar 1`] = `
Object {
  "error": Object {
    "message": "'predicate' is from wrong type object",
    "type": "Error",
  },
  "inputs": Array [
    Array [
      1,
    ],
    "foobarfoobar",
  ],
  "label": "errors are different",
  "ok": false,
  "ramdaError": Object {
    "message": "pred is not a function",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": "PENDING",
}
`;

exports[`brute force Function x => x < -2 | Number 2 1`] = `
Object {
  "error": Object {
    "message": "'iterable' is from wrong type number",
    "type": "Error",
  },
  "inputs": Array [
    [Function],
    2,
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": Array [],
  "result": "PENDING",
}
`;

exports[`brute force Function x => x < -2 | Promise  1`] = `
Object {
  "error": Object {
    "message": "'iterable' is from wrong type object",
    "type": "Error",
  },
  "inputs": Array [
    [Function],
    Promise {},
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": Array [],
  "result": "PENDING",
}
`;

exports[`brute force Function x => x < -2 | RegExp /foo/g 1`] = `
Object {
  "error": Object {
    "message": "'iterable' is from wrong type object",
    "type": "Error",
  },
  "inputs": Array [
    [Function],
    /foo/g,
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": Array [],
  "result": "PENDING",
}
`;

exports[`brute force Function x => x < 2 | Number 2 1`] = `
Object {
  "error": Object {
    "message": "'iterable' is from wrong type number",
    "type": "Error",
  },
  "inputs": Array [
    [Function],
    2,
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": Array [],
  "result": "PENDING",
}
`;

exports[`brute force Function x => x < 2 | Promise  1`] = `
Object {
  "error": Object {
    "message": "'iterable' is from wrong type object",
    "type": "Error",
  },
  "inputs": Array [
    [Function],
    Promise {},
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": Array [],
  "result": "PENDING",
}
`;

exports[`brute force Function x => x < 2 | RegExp /foo/g 1`] = `
Object {
  "error": Object {
    "message": "'iterable' is from wrong type object",
    "type": "Error",
  },
  "inputs": Array [
    [Function],
    /foo/g,
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": Array [],
  "result": "PENDING",
}
`;

exports[`brute force Function x => x > 2 | Number 2 1`] = `
Object {
  "error": Object {
    "message": "'iterable' is from wrong type number",
    "type": "Error",
  },
  "inputs": Array [
    [Function],
    2,
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": Array [],
  "result": "PENDING",
}
`;

exports[`brute force Function x => x > 2 | Promise  1`] = `
Object {
  "error": Object {
    "message": "'iterable' is from wrong type object",
    "type": "Error",
  },
  "inputs": Array [
    [Function],
    Promise {},
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": Array [],
  "result": "PENDING",
}
`;

exports[`brute force Function x => x > 2 | RegExp /foo/g 1`] = `
Object {
  "error": Object {
    "message": "'iterable' is from wrong type object",
    "type": "Error",
  },
  "inputs": Array [
    [Function],
    /foo/g,
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": Array [],
  "result": "PENDING",
}
`;

exports[`brute force Function x => x > 10 | Number 2 1`] = `
Object {
  "error": Object {
    "message": "'iterable' is from wrong type number",
    "type": "Error",
  },
  "inputs": Array [
    [Function],
    2,
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": Array [],
  "result": "PENDING",
}
`;

exports[`brute force Function x => x > 10 | Promise  1`] = `
Object {
  "error": Object {
    "message": "'iterable' is from wrong type object",
    "type": "Error",
  },
  "inputs": Array [
    [Function],
    Promise {},
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": Array [],
  "result": "PENDING",
}
`;

exports[`brute force Function x => x > 10 | RegExp /foo/g 1`] = `
Object {
  "error": Object {
    "message": "'iterable' is from wrong type object",
    "type": "Error",
  },
  "inputs": Array [
    [Function],
    /foo/g,
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": Array [],
  "result": "PENDING",
}
`;

exports[`brute force String  | Array [{}, 1, 2] 1`] = `
Object {
  "error": Object {
    "message": "'predicate' is from wrong type string",
    "type": "Error",
  },
  "inputs": Array [
    "",
    Array [
      Object {},
      "1",
      2,
    ],
  ],
  "label": "errors are different",
  "ok": false,
  "ramdaError": Object {
    "message": "pred is not a function",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": "PENDING",
}
`;

exports[`brute force String  | Array [1, 2, 3, 4, 5] 1`] = `
Object {
  "error": Object {
    "message": "'predicate' is from wrong type string",
    "type": "Error",
  },
  "inputs": Array [
    "",
    Array [
      1,
      2,
      3,
      4,
      5,
    ],
  ],
  "label": "errors are different",
  "ok": false,
  "ramdaError": Object {
    "message": "pred is not a function",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": "PENDING",
}
`;

exports[`brute force String  | Number 2 1`] = `
Object {
  "error": Object {
    "message": "'predicate' is from wrong type string",
    "type": "Error",
  },
  "inputs": Array [
    "",
    2,
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": Array [],
  "result": "PENDING",
}
`;

exports[`brute force String  | Promise  1`] = `
Object {
  "error": Object {
    "message": "'predicate' is from wrong type string",
    "type": "Error",
  },
  "inputs": Array [
    "",
    Promise {},
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": Array [],
  "result": "PENDING",
}
`;

exports[`brute force String  | RegExp /foo/g 1`] = `
Object {
  "error": Object {
    "message": "'predicate' is from wrong type string",
    "type": "Error",
  },
  "inputs": Array [
    "",
    /foo/g,
  ],
  "label": "Rambda should not throw",
  "ok": false,
  "ramdaError": "PENDING",
  "ramdaResult": Array [],
  "result": "PENDING",
}
`;

exports[`brute force String  | String foobar 1`] = `
Object {
  "error": Object {
    "message": "'predicate' is from wrong type string",
    "type": "Error",
  },
  "inputs": Array [
    "",
    "foobar",
  ],
  "label": "errors are different",
  "ok": false,
  "ramdaError": Object {
    "message": "pred is not a function",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": "PENDING",
}
`;

exports[`brute force String  | String foobarfoobar 1`] = `
Object {
  "error": Object {
    "message": "'predicate' is from wrong type string",
    "type": "Error",
  },
  "inputs": Array [
    "",
    "foobarfoobar",
  ],
  "label": "errors are different",
  "ok": false,
  "ramdaError": Object {
    "message": "pred is not a function",
    "type": "TypeError",
  },
  "ramdaResult": "PENDING",
  "result": "PENDING",
}
`;
